File: src/vision.pas
{$mode objfpc}{$H+}
{$assertions+}
{$overflowchecks on}
{$rangechecks on}
{$scopedenums on}

{ An implementation of Adam Milazzo's algorithm for tile visibility, which can be found at
  <http://www.adammil.net/blog/v125_Roguelike_Vision_Algorithms.html>.
}

unit Vision;

interface

uses Util;

type
    { Wld should provide the following API:
        function Wld.IsOpaque(Vec2): Boolean;          // whether a given point is solid
        function Wld.VisibilityDistance(Vec2): UInt32; // distance to origin
        procedure Wld.MarkVisible(Vec2);               // mark a given point as visible
    }
    generic VisionComputation<Wld> = object
    private
        type Octant = UInt8;
    private
        world: Wld;
        origin: Vec2;
        max_distance: UInt32;
        cur_octant: Octant;

        function Clamp(const a: UInt32): Int16; static;

        function AdjustForOctant(const p: Vec2): Vec2;
        function IsOpaque(const p: Vec2): Boolean;
        procedure MarkVisible(const p: Vec2);

        function SectorColumnTop   (const top:    Slope; const x: Int16): Int16;
        function SectorColumnBottom(const bottom: Slope; const x: Int16): Int16;
        procedure ScanColumn(
            var top, bottom: Slope;
            const x, max_y, min_y: Int16;
            var halt_now: Boolean
        );
        procedure ComputeSector(top, bottom: Slope; const min_x: Int16);
    public
        constructor Init(const world_: Wld; const origin_: Vec2; const max_distance_: UInt32);
        procedure Compute;
    end;

implementation

constructor VisionComputation.Init(const world_: Wld; const origin_: Vec2; const max_distance_: UInt32);
begin
    world := world_;
    origin := origin_;
    max_distance := max_distance_;
end;

{ The algorithm works by dividing up the world around the origin into eight triangular regions.
  Each region is bounded by one orthogonal line and one diagonal line, and coordinates are
  transformed such that vision computation in each octave can treat the orthogonal bounding line
  as 'y = 0' and and the diagonal bounding line is 'y = x'. }
function VisionComputation.AdjustForOctant(const p: Vec2): Vec2;
var x, y: Int16;
begin
    case cur_octant of
        0: begin x :=  p.x; y := -p.y; end;
        1: begin x :=  p.y; y := -p.x; end;
        2: begin x := -p.y; y := -p.x; end;
        3: begin x := -p.x; y := -p.y; end;
        4: begin x := -p.x; y :=  p.y; end;
        5: begin x := -p.y; y :=  p.x; end;
        6: begin x :=  p.y; y :=  p.x; end;
        7: begin x :=  p.x; y :=  p.y; end;
    else
        Assert(false, 'invalid octant');
    end;
    result.x := origin.x + x;
    result.y := origin.y + y;
end;

function VisionComputation.IsOpaque(const p: Vec2): Boolean;
begin
    result := world.IsOpaque(AdjustForOctant(p));
end;

procedure VisionComputation.MarkVisible(const p: Vec2);
begin
    world.MarkVisible(AdjustForOctant(p));
end;

function VisionComputation.Clamp(const a: UInt32): Int16;
const MAX_I16 = 32767;
begin
    if a > MAX_I16 then
        result := MAX_I16
    else
        result := Int16(a);
end;

function VisionComputation.SectorColumnTop(const top: Slope; const x: Int16): Int16;
var
    n, d: UInt16;
    ax: Int16;
begin
    n := top.numer; d := top.denom;
    if n = d then { slope = 1 }
    begin
        result := x;
        exit;
    end;

    result := ((2*x - 1) * n + d) div (2*d);

    if IsOpaque(MkVec2(x, result)) then
    begin
        if (top >= MkSlope(2*result + 1, 2*x))
           and not IsOpaque(MkVec2(x, result + 1))
        then
            Inc(result);
        exit;
    end;

    ax := 2*x;
    if IsOpaque(MkVec2(x + 1, result + 1)) then
        Inc(ax);
    if top > MkSlope(2*result + 1, ax) then
        Inc(result);
end;

function VisionComputation.SectorColumnBottom(const bottom: Slope; const x: Int16): Int16;
var
    n, d: UInt16;
begin
    n := bottom.numer; d := bottom.denom;
    if n = 0 then { slope = 0 }
    begin
        result := 0;
        exit;
    end;

    result := ((2*x - 1) * n + d) div (2*d);
    if (bottom >= MkSlope(2*result + 1, 2*x))
        and     IsOpaque(MkVec2(x, result))
        and not IsOpaque(MkVec2(x, result + 1))
    then
        Inc(result);
end;

procedure VisionComputation.ScanColumn(
    var top, bottom: Slope;
    const x, max_y, min_y: Int16;
    var halt_now: Boolean
);
var
    y: Int16;
    cell_loc: Vec2;
    cell_slope: Slope;
    is_top_cell, cell_is_opaque, last_cell_was_opaque: Boolean;
    tmp_slope: Slope;
begin
    Assert(not halt_now);
    is_top_cell := true;
    last_cell_was_opaque := true; // dummy value, never accessed in loop
                                  // when `is_top_cell` is set
    for y := max_y downto min_y do
    begin
        cell_loc := MkVec2(x, y);
        cell_slope := MkSlope(y, x);
        if (world.VisibilityDistance(cell_loc) > max_distance) then
            continue;
        // check if this cell should be visible
        if ((y <> max_y) or (top >= cell_slope)) and
           ((y <> min_y) or (cell_slope >= bottom))
        then
            MarkVisible(cell_loc);
        // no need to compute opacity on the last iteration
        if x = max_distance then
            continue;
        cell_is_opaque := IsOpaque(cell_loc);
        if (not is_top_cell) and (last_cell_was_opaque <> cell_is_opaque) then
        begin
            tmp_slope := MkSlope(2*y + 1, 2*x);
            if cell_is_opaque then
                if top > tmp_slope then
                    if y = min_y then
                    begin
                        bottom := tmp_slope;
                        break;
                    end
                    else
                        ComputeSector(top, tmp_slope, x + 1)
                else
                    if y = min_y then
                    begin
                        halt_now := true;
                        exit;
                    end
                    else
                    begin end
            else
                if bottom >= tmp_slope then
                begin
                    halt_now := true;
                    exit;
                end
                else
                    top := tmp_slope;
        end;
        is_top_cell := false;
        last_cell_was_opaque := cell_is_opaque;
    end;

    { this is true if `is_top_cell`, or if set normally }
    halt_now := last_cell_was_opaque;
end;

procedure VisionComputation.ComputeSector(top, bottom: Slope; const min_x: Int16);
var
    x: Int16;
    halt_now: Boolean;
begin
    halt_now := false;
    for x := min_x to Clamp(max_distance - 1) do
    begin
        ScanColumn(
            top, bottom,
            x,
            SectorColumnTop(top, x),
            SectorColumnBottom(bottom, x),
            halt_now
        );
        if halt_now then
            break;
    end;
end;

procedure VisionComputation.Compute;
var oct: Octant;
begin
    world.MarkVisible(origin);
    for oct := 0 to 7 do
    begin
        cur_octant := oct;
        ComputeSector(MkSlope(1, 1), MkSlope(0, 1), 1);
    end;
end;

end.

File: src/vision.zig
//! An implementation of Adam Milazzo's algorithm for tile visibility, which can be found at
//! <http://www.adammil.net/blog/v125_Roguelike_Vision_Algorithms.html>.

const std = @import("std");

const geometry = @import("geometry.zig");
const Vec2 = geometry.Vec2;
const Slope = geometry.Slope;

// The algorithm works by dividing up the world around the origin into eight triangular
// regions. Each region is bounded by one orthogonal line and one diagonal line, and
// coordinates are transformed such that vision computation in each octave can treat the
// orthogonal bounding line as 'y = 0' and and the diagonal bounding line is 'y = x'.
const Octant = u3;
fn adjustForOctant(origin: Vec2, o: Octant, p: Vec2) Vec2 {
    return origin.add(switch (o) {
        0 => Vec2.new(p.x, -p.y),
        1 => Vec2.new(p.y, -p.x),
        2 => Vec2.new(-p.y, -p.x),
        3 => Vec2.new(-p.x, -p.y),
        4 => Vec2.new(-p.x, p.y),
        5 => Vec2.new(-p.y, p.x),
        6 => Vec2.new(p.y, p.x),
        7 => Vec2.new(p.x, p.y),
    });
}

// World should provide the following API:
//    fn isOpaque(self, Vec2) bool       whether a given point is solid
//    fn magnitude(Vec2) u32             distance metric to origin
//    fn markVisible(self, Vec2) !void   mark a given point as visible
pub fn Visibility(comptime World: type, comptime Error: type) type {
    return struct {
        world: World,
        origin: Vec2,
        max_distance: u32,

        const Self = @This();

        fn isOpaque(self: *Self, o: Octant, p: Vec2) bool {
            return self.world.isOpaque(adjustForOctant(self.origin, o, p));
        }

        fn markVisible(self: *Self, o: Octant, p: Vec2) Error!void {
            return self.world.markVisible(adjustForOctant(self.origin, o, p));
        }

        fn vec(x: i32, y: i32) Vec2 {
            return Vec2.new(@intCast(i16, x), @intCast(i16, y));
        }

        fn slope(x: i32, y: i32) Slope {
            return Slope.new(@intCast(u16, x), @intCast(u16, y));
        }

        /// Recursively add visible points in the sector bounded by the lines given by
        /// `top` and `bottom` and whose relative x coordinate is at least `min_x`.
        fn computeSector(
            self: *Self,
            oct: Octant,
            param_top: Slope,
            param_bottom: Slope,
            min_x: i32,
            max_x: i32,
        ) Error!void {
            var top = param_top;
            var bottom = param_bottom;
            var x = min_x;
            while (x < max_x) : (x += 1) {
                const column_top_y = if (top.numer == top.denom)
                    x
                else blk: {
                    const n = top.numer;
                    const d = top.denom;
                    var top_y = @intCast(i32, ((2 * @intCast(u32, x) - 1) * n + d) / (2 * d));
                    if (self.isOpaque(oct, vec(x, top_y))) {
                        if (top.ge(slope(2 * top_y + 1, 2 * x)) and
                            !self.isOpaque(oct, vec(x, top_y + 1)))
                            top_y += 1;
                    } else {
                        var ax = 2 * x;
                        if (self.isOpaque(oct, vec(x + 1, top_y + 1)))
                            ax += 1;
                        if (top.gt(slope(top_y + 1, ax)))
                            top_y += 1;
                    }
                    break :blk top_y;
                };
                const column_bottom_y = if (bottom.numer == 0)
                    0
                else blk: {
                    const n = bottom.numer;
                    const d = bottom.denom;
                    var bottom_y = @intCast(i32, ((2 * @intCast(u32, x) - 1) * n + d) / (2 * d));
                    if (bottom.ge(slope(2 * bottom_y + 1, 2 * x)) and
                        self.isOpaque(oct, vec(x, bottom_y)) and
                        !self.isOpaque(oct, vec(x, bottom_y + 1)))
                        bottom_y += 1;
                    break :blk bottom_y;
                };
                var last_cell_was_opaque: ?bool = null;
                var y = column_top_y;
                while (y >= column_bottom_y) : (y -= 1) {
                    const cell_loc = vec(x, y);
                    if (World.magnitude(cell_loc) > self.max_distance) continue;
                    {
                        const cell_slope = slope(y, x);
                        // check if this cell should be visible
                        if ((y != column_top_y or top.ge(cell_slope)) and
                            (y != column_bottom_y or cell_slope.ge(bottom)))
                            try self.markVisible(oct, cell_loc);
                    }

                    const cell_is_opaque = self.isOpaque(oct, cell_loc);
                    defer last_cell_was_opaque = cell_is_opaque;
                    if (last_cell_was_opaque != !cell_is_opaque) continue;

                    const mid_slope = slope(2 * y + 1, 2 * x);
                    if (cell_is_opaque)
                        if (top.gt(mid_slope))
                            if (y == column_bottom_y) {
                                bottom = mid_slope;
                                break;
                            } else try self.computeSector(oct, top, mid_slope, x + 1, max_x)
                        else if (y == column_bottom_y)
                            return
                        else {}
                    else if (bottom.ge(mid_slope))
                        return
                    else
                        top = mid_slope;
                }
                if (last_cell_was_opaque != false)
                    break;
            }
        }

        pub fn compute(self: *Self) Error!void {
            try self.world.markVisible(self.origin);
            const max_x = @intCast(i32, self.max_distance);
            var oct: Octant = 0;
            while (true) {
                try self.computeSector(oct, Slope.new(1, 1), Slope.new(0, 1), 1, max_x);
                if (oct == 7) break else oct += 1;
            }
        }
    };
}

test "compilation" {
    std.testing.refAllDecls(@This());
    std.testing.refAllDecls(Visibility(struct {
        const Self = @This();
        fn isOpaque(self: Self, at: Vec2) bool {
            _ = self;
            _ = at;
            @panic("unimplemented");
        }
        fn magnitude(at: Vec2) u32 {
            _ = at;
            @panic("unimplemented");
        }
        fn markVisible(self: Self, at: Vec2) void {
            _ = self;
            _ = at;
            @panic("unimplemented");
        }
    }));
}
